### PR TITLE
Story/vspc 20

### PR DESCRIPTION
**Pull request by Neehaarika**

### CODE REVIEW

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
there are test failures.

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SequenceManager.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    private SequenceRepository sequenceRepo;
+
+    @Override
+    public ISequence storeSequence(String moduleId, SequenceForm sequenceForm) {
+        List<ISlide> slides = new ArrayList<>();
+        for(String slideId : sequenceForm.getOrderedSlides()) {
+            slides.add(slideManager.getSlide(slideId));
+        }
+        ISequence sequence = sequenceFactory.createSequence(moduleManager.getModule(moduleId), sequenceForm, slides);
+        sequenceRepo.save((Sequence) sequence);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/SequenceManager.java]: you probably want to return the sequence object that ```save``` returns. I believe the repo creates a new sequence object with id and does not set the id for the passed object.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+import edu.asu.diging.vspace.web.staff.forms.SequenceForm;
+
+public class SequenceFactoryTest {
+
+    @InjectMocks
+    private SequenceFactory sequenceFactoryToTest = new SequenceFactory();
+    
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceFactoryTest.java]: since you are not using any mocks in this test class you can remove this line and the ```injectMocks``` annotation.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    @Before
+    public void init() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void test_getSequence_idExists() {
+        Sequence newSequence = new Sequence();
+        newSequence.setId("sequence1");
+        Optional<Sequence> mockSequence = Optional.of(newSequence);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: this is not a mocked sequence. just a sequence. the variable name is confusing.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        sequenceForm.setOrderedSlides(slideIds);
+        
+        ISlide slide1 = new Slide();
+        slide1.setId("slide1");       
+        ISlide slide2 = new Slide();
+        slide2.setId("slide2");
+ 
+        List<ISlide> slides = new ArrayList<>();
+        slides.add(slide1);
+        slides.add(slide2);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: I believe there is a ```Arrays.asList()``` method that just takes several objects and returns a list of them. This would turn these three lines into a one-liner.

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        Mockito.when(sequenceFactory.createSequence(module, sequenceForm, slides)).thenReturn(newSequence);
+        Mockito.when(mockSequenceRepo.save(newSequence)).thenReturn(newSequence);
+        
+        ISequence actualSequence = sequenceManagerToTest.storeSequence(module.getId(), sequenceForm);
+
+        Assert.assertEquals(actualSequence.getModule().getId(), module.getId());
+        Assert.assertEquals(actualSequence.getName(), sequenceForm.getName());
+        Assert.assertEquals(actualSequence.getDescription(), sequenceForm.getDescription());
+        Assert.assertEquals(actualSequence.getSlides(), slides);
+        Mockito.verify(mockSequenceRepo).save(newSequence);
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: after a sequence has been stored, it should have an id.

**Comment by jdamerow** [comment]
Jenkins, test this please.

**Comment by jdamerow** [comment]
everything else looks good!

**Comment by jdamerow** [comment]
Jenkins, test this please.

File: vspace/.gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 /target/
 /Users/
 /.settings/
-.DS_Store
+/.DS_Store
```
Comment (jdamerow) [inline vspace/.gitignore]: why this change?

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceFactoryTest.java (hunk 2 above)
Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceFactoryTest.java]: ?

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    @Test
+    public void test_getSequence_idExists() {
+        Sequence newSequence = new Sequence();
+        newSequence.setId("sequence1");
+        Optional<Sequence> sequence = Optional.of(newSequence);
+        Mockito.when(mockSequenceRepo.findById(newSequence.getId())).thenReturn(sequence);
+
+        ISequence actualSequence = sequenceManagerToTest.getSequence(newSequence.getId());
+        Assert.assertEquals(newSequence.getId(), actualSequence.getId());
+        Assert.assertEquals(newSequence.getName(), actualSequence.getName());
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: you are not setting the name of the sequence

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+        sequenceForm.setDescription("sample description");            
+        List<String> slideIds = Arrays.asList("slide1", "slide2");
+        sequenceForm.setOrderedSlides(slideIds);
+        
+        List<ISlide> slidesList = Arrays.asList(new Slide(), new Slide());
+        slidesList.get(0).setId("slide1");
+        slidesList.get(1).setId("slide2");
+            
+        Sequence newSequence = new Sequence();
+        newSequence.setId("seq01");
```
Comment (jdamerow) [inline vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: when ```storeSequence``` is called, the sequence does not have an id yet; the returned sequence has one (that's how you know it was stored). So  ```newSequence``` should not have an id. The sequence object returned by the repo, however, has.

**Comment by jdamerow** [comment]
tests seem to be failing

File: vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java (hunk 8 above)
Comment (Neehaarika) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: but we are the ones who are telling mockito what to return when save method is hit on the repo right? it is returning newSequence object and Assert.assertNotNull(actualSequence.getId()); is failing as it returns null.

Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: but let's say someone changes the code so that it checks if the ID is set after the factory is called and then returns null because it assumes the object is not a new object, then your test would fail although the code would still be correct. Your test case needs to mock what is actually happening (id = null before save, != null after save). Although, you probably want to test that the id the returned object has is the id that the save method returns, not a different one. Otherwise, if someone would change the code to set the id manually after save is called, your test would pass although the code would be wrong.

Comment (jdamerow) [reply vspace/src/test/java/edu/asu/diging/vspace/core/services/impl/SequenceManagerTest.java]: the easiest would probably to create two sequence objects, one returned by the factory, one returned by the repo. the other option is to use ```doAnswer``` as described here (with actually exactly our case): https://wahz.blogspot.com/2012/10/mockito-modifying-parameters.html

**Comment by jdamerow** [comment]
Jenkins, test this please.
